sdk/servicebus/Microsoft.Azure.ServiceBus/src/Primitives/ConcurrentExpiringSet.cs

```diff
@@ -40,28 +52,38 @@ void ScheduleCleanup()
                 }
 
                 this.cleanupScheduled = true;
-                Task.Run(async () => await this.CollectExpiredEntriesAsync().ConfigureAwait(false));
+                _ = this.CollectExpiredEntriesAsync(token);
             }
         }
 
-        async Task CollectExpiredEntriesAsync()
+        async Task CollectExpiredEntriesAsync(CancellationToken token)
         {
-            await Task.Delay(delayBetweenCleanups);
-
-            lock (this.cleanupSynObject)
+            try
             {
-                this.cleanupScheduled = false;
+                await Task.Delay(delayBetweenCleanups, token).ConfigureAwait(false);
+            }
+            catch (OperationCanceledException)
+            {
+                return;
+            }
+            finally
+            {
+                lock (this.cleanupSynObject)
+                {
+                    this.cleanupScheduled = false;
+                }
             }
 
-            foreach (var key in this.dictionary.Keys)
+            foreach (var kvp in this.dictionary)
             {
-                if (DateTime.UtcNow > this.dictionary[key])
+                var expiration = kvp.Value;
+                if (DateTime.UtcNow > expiration)
                 {
-                    this.dictionary.TryRemove(key, out _);
+                    this.dictionaryAsCollection.Remove(kvp);
                 }
             }
 
-            this.ScheduleCleanup();
+            this.ScheduleCleanup(token);
```

Comment:
What happens if the token is canceled?

Reply:
It will cancel everything and not throw see try catch further down. Are you hinting that you'd like an realy exit before acquiring the lock?